In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
path = r'/Users/vanessanunezpenas/Documents/Data Analytics Program - CareerFoundry/II Data Immersion 1 - Preparing and Analyzing Data/A4 - Python Fundamentals/2024.01 Instacart Basket Analysis'
path

'/Users/vanessanunezpenas/Documents/Data Analytics Program - CareerFoundry/II Data Immersion 1 - Preparing and Analyzing Data/A4 - Python Fundamentals/2024.01 Instacart Basket Analysis'

In [3]:
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

In [4]:
# Create a subset of the dataframe with the first one million rows

df = ords_prods_merge[:1000000]

In [5]:
df.shape

(1000000, 14)

## If-Statements with User-Defined Functions

In [6]:
# Define a function to flag products according to price

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [7]:
# Apply the function

df['price_range'] = df.apply(price_label, axis=1)

/var/folders/ty/6wr65d8n36sc80bglktwzwyw0000gn/T/ipykernel_76205/93659844.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [8]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [9]:
df['prices'].max()

14.8

## If-Statements with the loc() Function

In [10]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [11]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [12]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [13]:
ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21861997
Low-range product     10126366
High-range product      417678
Name: count, dtype: int64

## If-Statements with For-Loops

In [14]:
ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204404
1    5660456
6    4496635
2    4213986
5    4205906
3    3840701
4    3783953
Name: count, dtype: int64

In [15]:
result = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [16]:
ords_prods_merge['busiest_day'] = result

In [17]:
ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22417684
Busiest day        6204404
Least busy         3783953
Name: count, dtype: int64

## Answering some key Questions

In [18]:
# checking the new columns “price_label” and “busiest_day”

ords_prods_merge.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', '_merge', 'price_range_loc', 'busiest_day'],
      dtype='object')

In [22]:
# create a new the 'busiest_days' column

result_2 = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value == 1:
    result_2.append("Busiest days")
  elif value == 4 or value == 3:
    result_2.append("Slowest days")
  else:
    result_2.append("Regularly busy")

In [23]:
ords_prods_merge['busiest_days'] = result_2

In [24]:
# Check the values of this new column for accuracy

ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12916527
Busiest days      11864860
Slowest days       7624654
Name: count, dtype: int64

In [25]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day    busiest_days  
0  Regularly busy  Regularly busy  
1  Regularly busy    Slowest days  
2  Regularly busy    Slowest days  
3      Least busy    Slowest days  
4      Least busy    Slowest days

After flagging the the two busiest days (Saturday and Sunday) and the two least busiest days (Wednesday and Tuesday), the most frequent value is "Regularly busy" (12,916,527), followed by "Busiest days" (11,864,860) and "Slowest days" (7,624,654). 

In [26]:
# identify the busiest hours of the day

ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761862
11    2736242
14    2689251
15    2662246
13    2661057
12    2618613
16    2535313
9     2454273
17    2087721
8     1718182
18    1636566
19    1258352
20     976182
7      891082
21     795658
22     634247
23     402335
6      290500
0      218786
1      115706
5       87964
2       69377
4       53245
3       51281
Name: count, dtype: int64

We will use an If-Statement with For-Loop to label busiest_period_of_day:
    - 'Most orders': >= 2 millionen orders/hour;
    - 'Average orders': 1-2 millionen order/hour;
    - 'Fewest orders': <= 1 million orders/hour

In [29]:
hour_counts = ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

In [30]:
result_3 = []

for hour in ords_prods_merge['order_hour_of_day']:
    count = hour_counts[hour]
    if count >= 2000000:
        result_3.append("Most orders")
    elif count <= 1000000:
        result_3.append("Fewest orders")
    else:
        result_3.append("Average orders")

In [31]:
ords_prods_merge['busiest_period_of_day'] = result_3

In [32]:
# Frequency of the new 'busiest_period_of_day'

ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       23206578
Average orders     4613100
Fewest orders      4586363
Name: count, dtype: int64

In [33]:
# Exporting the dataframe as a pickle file

ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_derived.pkl'))